cjworkbench/i18n/trans.py

```diff
@@ -90,22 +79,67 @@ def localize_html(
     
     HTML is escaped in the message, as well as in arguments and tag attributes.
     """
-    if locale_id != default_locale:
-        try:
-            return _get_translations(locale_id).get_and_format_html_message(
-                message_id, context=context, arguments=arguments, tags=tags
-            )
-        except ICUError as err:
-            logger.exception(
-                f"Error in po file for locale {locale_id} and message {message_id}: {err}"
-            )
-        except KeyError as err:
-            pass
-    return _get_translations(default_locale).get_and_format_html_message(
-        message_id, context=context, arguments=arguments, tags=tags
+    return MessageLocalizer.for_application_messages(locale_id).localize_html(
+        message_id, arguments=arguments, tags=tags, context=context
     )
 
 
+class MessageLocalizer:
+    def __init__(self, locale_id: str, catalog: Catalog, default_catalog: Catalog):
+        self.locale_id = locale_id
+        self.formatter = MessageFormatter(locale_id, catalog)
+        self.default_formatter = MessageFormatter(default_locale, default_catalog)
+
+    def localize(self, message_id, arguments={}) -> str:
+        if self.locale_id != default_locale:
+            try:
+                return self.formatter.get_and_format_message(
+                    message_id, arguments=arguments
+                )
+            except ICUError as err:
+                logger.exception(
+                    f"Error in po file for locale {self.locale_id} and message {message_id}: {err}"
+                )
+            except KeyError as err:
+                pass
+        return self.default_formatter.get_and_format_message(
+            message_id, arguments=arguments
+        )
+
+    def localize_html(self, message_id, *, context, arguments, tags) -> str:
+        if self.locale_id != default_locale:
+            try:
+                return self.formatter.get_and_format_html_message(
+                    message_id, context=context, arguments=arguments, tags=tags
+                )
+            except ICUError as err:
+                logger.exception(
+                    f"Error in po file for locale {self.locale_id} and message {message_id}: {err}"
+                )
+            except KeyError as err:
+                pass
+        return self.default_formatter.get_and_format_html_message(
+            message_id, context=context, arguments=arguments, tags=tags
+        )
+
+    @classmethod
+    @lru_cache(maxsize=len(supported_locales))
+    def for_application_messages(cls, locale_id: str):
+        """Return a `MessageLocalizer` for the application internal messages in the given locale.
+
+        Uses lru_cache in order to parse the message catalogs only once per locale.
+        """
+        return cls(locale_id, load_catalog(locale_id), load_catalog(default_locale))
+
+    @classmethod
+    def for_module(cls, module_zip: ModuleZipfile, locale_id: str):
+        return cls(
```

Comment:
We should cache here. For two reasons: A) it seems wrong to construct a new `MessageLocalizer` every time we call `for_module()`; and B) `cjwstate` shouldn't depend on Babel.

Let's use a class-member dictionary cache, keyed by module. `self._module_localizers: Dict[str, Dict[str, MessageLocalizer]] = weakref.WeakKeyDictionary()`. Use [WeakKeyDictionary](https://docs.python.org/3/library/weakref.html#weakref.WeakKeyDictionary): it won't hold a reference to the ModuleZipfile, and it'll delete catalogs belonging to deleted ModuleZipfiles. Remember to use a `threading.Lock` when accessing the dictionary.